### ISSUE TITLE
Aggregate root design

### ISSUE THREAD

**Issue by johncollinson2001** [dibley1973/OpenRMS#32]
Location and location type... Are these 2 aggregate roots?

I think they are, it's widely accepted that aggregate roots should be small and concise, but my understanding is because the identity of a location type lives across multiple locations, it is not aggregated under a location.

Any thoughts?

**Comment by jadjare** [comment]
Is Location Type even an entity?  The word "Type", in my mind, implies either Inheritance or a ValueObject.  In the physical world, at least, you always have a type of something.  Something being the Entity and Type implying a variation of.

At this point in development I'm thinking it should NOT be treated as an entity / aggregate... If, in fact, a location type is an aggregate it will reveal itself.


Ah, OK.  I've just thought, are you thinking of a Location Type as an attribute that's attached to a Location? And location types will be managed separately?

**Comment by johncollinson2001** [comment]
Yes exactly as per your last sentence. To be honest this is the database driven developer getting the better of me - I've got a lookup of location types in my head!

However a location type shared across locations, should it be an entity with an id in the domain?

**Comment by jadjare** [comment]
Hmm, grabbed my Vaughn DDD book for a little light reading!!

In the meantime and not specific to this problem...
...I'm thinking that an attribute, which at any point can have new ones defined, implies a piece of reference data simply used to group/order data.  That is unless the attribute has a bunch of flags associated to it and in that way can be used to describe intent.  The system can then behave according to the intent.  If this is the case it implies there's a finite number of behaviours.  This finite number of behaviours is then in essence our types.  Thus we only care about types if they affect behaviour.  Short of a very dynamic model changes in behaviour will require a code change somewhere.  I guess I'm thinking if, and only if at this stage, type affects behaviour it's best modelled through Inheritance, else for it's just a Value which holds no meaning (in the bounded context)... Note this Value maybe an ID pointer that in another context may hold meaning.

Anyway that's just a brain dump to stimulate further discussion.  Hopefully I will get a chance to read up more later tonight.

**Comment by johncollinson2001** [comment]
Sort of followed along there! When you say modelled through inheritence do you mean base type of LocationType, sub types of Store, DC, etc?

Going by the acceptance criteria for the Location Type issue i'm working on - #2 - it should have an identifier which means something to the business, the type need to be maintained, Does all this stand true? If so then it's an entity in some shape or form. However I'm not sure if as far as location types go, it's nothing more than an enum.

Some interesting thoughts on the topic here http://stackoverflow.com/questions/12870650/should-lookup-values-be-modeled-as-aggregate-roots

But, I'm still not 100% clear on the best way to go about it... Can we all discuss whether the acceptance criteria for #2 is correct? If so then I think it answers the question.

**Comment by jadjare** [comment]
I perhaps should have read issue #2 before commenting, but that would be obvious!

I see it's the 2nd ever issue added in the project and actually was created before #3 add location.

To be honest I don't think the requirement is appropriate at this stage.  Marking a location with a Type serves no purpose until it's needed to differentiate locations.  With the basics we've got at the moment I don't think there's any call to differentiate one location from the next.  That requirement may or may not come up when we tackle Inventory Management or the other subsequent contexts.

**If you don't want to read any more of my waffle you can stop here, if not proceed at you peril!**
I also suspect that when we look at Location Types there will be a finite number and they won't need to be user definable.  This is because I think it's almost certain that  we'll only be interested in distinguishing one location from another in order to implement business rules. 
I guess in my head I'm distinguishing between a location type and an attribute, the latter being a more general purpose piece of reference data.  It may come up that we need to assign locations attributes and these are for the purpose of finding, grouping and sorting locations.  

In the instance of location types, if they are indeed finite and only possible to be added to by developers, then I think they'll sit as Value Types, even if maintained in the database.  I think this is line with commentary in the linked stack overflow post.

In the case of my attribute distinction, I guess the question will be if the user assigns the attribute of "City Centre" to 100 locations, and then later decides to edit the attribute to "Town Centre" is the intent to redefine all 100 locations.  If it is, I guess the attribute is an entity (with it's own aggregate root), if it's not then I guess it's a ValueObject.

I do also recall hearing somewhere (I think Vladimir Khorikov's DDD in Practice course) that it's perfectly acceptable for something to exist as an Entity in one context and a ValueObject in another. So it may be worth bearing that in mind.

Anyway enough of the waffle, as I said earlier, I personally don't think the requirement should be implemented at this stage.  If others agree this will be a moot point, for now at least.

**Comment by johncollinson2001** [comment]
Agreed, implement at the point of need. It's not really needed right now...
